typing/typecore.ml

```diff
@@ -1950,13 +1950,39 @@ struct
 
   type sd = Static | Dynamic
 
-  let rec classify_expression : Typedtree.expression -> sd =
-    fun exp -> match exp.exp_desc with
-      | Texp_let (_, _, e)
+  let classify_expression : Typedtree.expression -> sd =
+    (* We need to keep track of the size of expressions
+       bound by local declarations, to be able to predict
+       the size of variables. Compare:
+
+         let rec r =
+           let y = fun () -> r ()
+           in y
+
+       and
+
+         let rec r =
+           let y = if Random.bool () then ignore else fun () -> r ()
+           in y
+
+      In both cases the final adress of `r` must be known before `y` is compiled,
+      and this is only possible if `r` has a statically-known size.
+
+      The first definition can be allowed (`y` has a statically-known
+      size) but the second one is unsound (`y` has no statically-known size).
+    *)
+    let rec classify_expression env e = match e.exp_desc with
+      (* binding and variable cases *)
+      | Texp_let (rec_flag, vb, e) ->
+          let env = classify_value_bindings rec_flag env vb in
+          classify_expression env e
+      | Texp_ident (path, _, _) ->
+          classify_path env path
+      (* easy cases *)
       | Texp_letmodule (_, _, _, e)
       | Texp_sequence (_, e)
-      | Texp_letexception (_, e) -> classify_expression e
-      | Texp_ident _
+      | Texp_letexception (_, e) ->
+          classify_expression env e
```

Comment:
There are several changes here that are purely about style (indentation, etc.).  Perhaps they could be put into a separate commit.